nhais-adaptor/edifact/incoming/parser/deserialiser.py

```diff
@@ -0,0 +1,100 @@
+import edifact.incoming.parser.creators as creators
+from edifact.incoming.models.message import MessageSegment
+from edifact.incoming.models.interchange import Interchange
+
+INTERCHANGE_HEADER_KEY = "UNB"
+MESSAGE_HEADER_KEY = "UNH"
+MESSAGE_BEGINNING_KEY = "BGM"
+MESSAGE_REGISTRATION_KEY = "S01"
+MESSAGE_PATIENT_KEY = "S02"
+MESSAGE_TRAILER_KEY = "UNT"
+INTERCHANGE_TRAILER_KEY = "UNZ"
+
+
+def extract_relevant_lines(original_dict, starting_pos, trigger_key):
+    """
+    From the original dict generate a smaller dict just containing the relevant lines based upon the trigger key
+    will keep looping till the terminating key is found in the terminating config
+    :param original_dict: The original larger dictionary
+    :param starting_pos: The starting position to start the loop from.
+    This is to prevent starting the loop from the start each time and be slightly more efficient
+    :param trigger_key: The trigger key for this section that will be used to find what the
+    terminating key for the section is
+    :return: A smaller dictionary with just the relevant lines for the section
+    """
+    terminating_config = {
+        INTERCHANGE_HEADER_KEY: [MESSAGE_HEADER_KEY],
+        MESSAGE_BEGINNING_KEY: [MESSAGE_REGISTRATION_KEY],

```

Comment:
Isn't this quite tightly tied to the structure of the patient registration message we're parsing? i.e. The message beginning section is presumably common across messages, so it won't always be followed by a message registration key (as that is presumably specific to patient registration messages).

Reply:
Yes, that's right. So for example an `SO2` section may not exist in the response. So the terminator for `SO1` can be an `SO2` or a `UNT` (message trailer). This is why I've added the terminating keys as a List so it can be added to easily. I'm hoping working on another example may unearth more. But for now this is the structure I've come across.